### PR TITLE
Upgrade to MongoDB 7.0.4

### PR DESCRIPTION
**Pull request by Xiangs18**

### CODE REVIEW

**Comment by Xiangs18** [comment]
@MrCreosote 
https://github.com/kbase/workspace_deluxe/actions/runs/7978938773/job/21785169860?pr=715#step:6:1881
Have you seen errors like this before? There are 5 of them failed for the same reason.

**Comment by MrCreosote** [comment]
It looks like the authcontroller isn't starting up properly. Does everything work with Mongo 3? If not we may need to update the auth jar once it's released before the tests will pass

**Comment by Xiangs18** [comment]
Everything works with Mongo 3.

**Comment by Xiangs18** [comment]
wait, are you saying we should use new auth2 jar to test?

**Comment by MrCreosote** [comment]
> If tests pass w/ mongo 3 and fail with 7 it looks like we'll need the new auth jar, yeah. Check the auth server logs to be sure

**Comment by MrCreosote** [comment]
I'm not sure why any changes are needed - I got the WS tests to pass with the new auth jar. 

1. The test classes and main classes should not be in the same package to ensure only the public API are tested. 
2. The authcontroller is for tests, hence it's in the test package. It's definitely not for use in non-test code

**Comment by Xiangs18** [comment]
May I know how did you import AuthController class from the new auth jar?

**Comment by MrCreosote** [comment]
I didn't have to change the imports at all from what I recall - unfortunately I deleted the changes once I confirmed it worked. Let me know if you can't get things working and I'll set things up again

**Comment by Xiangs18** [comment]
Here is what I am struggling:
1. `./gradlew jar` in auth2 to create the new auth2 jar 
3.  Try to use new auth2 jar in the ws repo, but cannot `import us.kbase.test.auth2.authcontroller.AuthController;`

**Comment by MrCreosote** [comment]
You need to use the fastTestJar target: https://github.com/kbase/auth2/blob/develop/build.gradle#L111

**Comment by Xiangs18** [comment]
Updated as the last commit suggested!

**Review by MrCreosote — COMMENTED** [review]
Looks pretty good, just some minor comments

File: build.xml [hunk 1]
```diff
@@ -101,7 +104,7 @@
     <include name="apache_commons/commons-collections-3.2.1.jar"/>
     <include name="jsemver/java-semver-0.9.0.jar"/>
     <include name="equalsverifier/equalsverifier-3.1.10.jar"/>
-    <include name="kbase/auth2/kbase-auth2test-0.2.4.jar"/>
+    <include name="kbase/auth2/kbase-auth2-test-fat-0.6.1.jar"/>
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply build.xml:107]: Could you please explain the term `shadow other jars` here more? I don't think I understand it 100%. Thanks!

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -56,6 +56,12 @@ jobs:
             minio: '2019-05-23T00-29-34Z'
             wired_tiger: 'false'
             ant_test: 'test_quick_coverage'
+          # Mongo7 setup
+          - java: '11'
+            mongo: 'mongodb-linux-x86_64-ubuntu2204-7.0.4'
+            minio: '2019-05-23T00-29-34Z'
+            wired_tiger: 'false'
+            ant_test: 'test_quick_coverage'
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply .github/workflows/test.yml]: 👍

File: build.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xiangs18) [reply build.xml]: 👍

File: src/us/kbase/workspace/test/WorkspaceMongoIndex.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import org.bson.Document;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class WorkspaceMongoIndex {
+
+    public static Set<Document> getAndNormalizeIndexes(final MongoDatabase db, final String collectionName) {
+        final Set<Document> indexes = new HashSet<>();
+        for (Document index: db.getCollection(collectionName).listIndexes()) {
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply src/us/kbase/workspace/test/WorkspaceMongoIndex.java]: 👍

File: src/us/kbase/workspace/test/WorkspaceMongoIndex.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+            // In MongoDB 4.4, the listIndexes and the mongo shell helper method db.collection.getAndNormalizeIndexes()
```
[the remark this answers is not in the document]
Comment (Xiangs18) [reply src/us/kbase/workspace/test/WorkspaceMongoIndex.java]: I will be more careful when using `Replace All` next time!

[the remark this answers is not in the document]
Comment (Xiangs18) [reply src/us/kbase/workspace/test/WorkspaceMongoIndex.java]: 👍

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply .github/workflows/test.yml]: wired_tiger should be false for the top setup

[the remark this answers is not in the document]
Comment (MrCreosote) [reply .github/workflows/test.yml]: :+1:

File: build.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply build.xml]: :+1:

File: src/us/kbase/workspace/test/WorkspaceMongoIndex.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/us/kbase/workspace/test/WorkspaceMongoIndex.java]: :+1:

File: src/us/kbase/workspace/test/WorkspaceMongoIndex.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/us/kbase/workspace/test/WorkspaceMongoIndex.java]: :+1:

**Comment by MrCreosote** [comment]
Looks good, but I just realized it needs release notes & a version bump to 0.14.3

**Pull request merged**